### PR TITLE
support topics assignation (to and condition)

### PR DESCRIPTION
**Pull request by juliangut**
https://firebase.google.com/docs/cloud-messaging/http-server-ref#table1

### CODE REVIEW

File: library/Gcm/Message.php [hunk 1]
```diff
@@ -419,6 +486,22 @@ public function getDryRun()
...
+            $json['to'] = (!preg_match('!^/topics/!', $to) ? '/topics/' : '') . $to;
+        } elseif ($this->topics) {
+            $conditions = array_map(
+                function ($topic) {
+                    return $topic . (!preg_match('/ in topics$/', $topic) ? ' in topics' : '');
+                },
+                $this->topics
+            );
+
+            $json['condition'] = implode(' || ', $conditions);
```
Comment (mwillbanks) [inline library/Gcm/Message.php]: Right now the condition is hard coded to an OR condition vs supporting both the && and the || conditions.  Think you might be able to support both styles of conditions?

Comment (juliangut) [inline library/Gcm/Message.php]: I've added a new field: conditionOperator set to `||` by default.

File: library/Gcm/Message.php [hunk 2]
```diff
@@ -449,4 +548,37 @@ public function toJson()
 
         return Json::encode($json);
     }
+
+    /**
+     * Get message target parameters for JSON.
+     *
+     * @return array
+     */
+    protected function getTarget()
```
Comment (juliangut) [inline library/Gcm/Message.php]: Message target logic:

1. If only ONE topic provided it is used as "to" parameter (registrationIds are discarded)
2. If only ONE registrationIds provided it is used as "to" parameter
3. If more than one registrationIds it they are set as "registration_ids" parameter
4. If more than one topic are provided they are combined with conditionOperator to form "conditoin" parameter

File: library/Gcm/Message.php [hunk 3]
```diff
@@ -449,4 +548,37 @@ public function toJson()
...
+        } elseif ($this->registrationIds) {
+            $target['registration_ids'] = $this->registrationIds;
+        }
+
+        if (count($this->topics) > 1) {
+            $conditions = array_map(
+                function ($topic) {
+                    return $topic . (!preg_match('/ in topics$/', $topic) ? ' in topics' : '');
+                },
+                $this->topics
```
[the remark this answers is not in the document]
Comment (mustafa-taheri) [reply library/Gcm/Message.php]: Sorry to bother you @juliangut but can you please explain this with an example how can I send a notification to a single topic. I am a bit confused...

[the remark this answers is not in the document]
Comment (weierophinney) [reply library/Gcm/Message.php]: > That means maximum three topics allowed but I can't decide if just array_slice($this->topics, 0, 3) or raise an error when assigning more than 3 topics

Raise an error if more than 3 topics are assigned. Most likely, the developer did not know about this limitation, and if we were to silently truncate, they may wonder why topics are not being sent.

File: tests/Gcm/MessageTest.php [hunk 4]
```diff
@@ -59,7 +64,34 @@ public function testExpectedRegistrationIdBehavior()
...
+        $this->m->clearTopics();
+        self::assertEquals($this->m->getTopics(), []);
+        self::assertNotContains('"to"', $this->m->toJson());
+        self::assertNotContains('condition', $this->m->toJson());
+        $this->m->addTopic('general');
+        self::assertEquals($this->m->getTopics(), ['general']);
+        self::assertContains('"to"', $this->m->toJson());
+        self::assertContains('\/topics\/general', $this->m->toJson());
+        self::assertNotContains('condition', $this->m->toJson());
     }
```
Comment (weierophinney) [inline tests/Gcm/MessageTest.php]: Please break this into multiple tests:

- A test for empty topics by default.
- A test for behavior when valid topics are added, including an assertion around the default conditional
- A test for changing the conditional
- A test detailing clearing topics
- A test detailing what happens when too many topics are added

**Comment by weierophinney** [comment]
Closing due to:

- lack of feedback from author (maintainers have several requests that have not yet been addressed)
- needs rebase

If anybody reading this would like to see this patch land in a release, please re-submit, with additional changes addressing the feedback provided.

**Pull request closed without merging**